packages/shared/scope-hoisting/src/link.js

```diff
@@ -337,32 +337,28 @@ export function link({
           }
         } else {
           if (mod.meta.id && assets.get(assertString(mod.meta.id))) {
+            let name = assertString(mod.meta.exportsIdentifier);
             // Replace with nothing if the require call's result is not used.
             if (!isUnusedValue(path)) {
-              let name = assertString(mod.meta.exportsIdentifier);
               node = t.identifier(replacements.get(name) || name);
+            }
```

Comment:
That we only run the code to insert `__esModule` if the `require` call is removed did not look related to me. Was there any reason for it?

Reply:
It was only inserted if the call was *not* removed (`mod` is truthy) and if the call is `!isUnused(...)`?

Reply:
Exactly. Now it is always inserted in case `mod` is truthy, independent of what `!isUnused(...)` returns.

Reply:
For that, we should use `!isUnusedValue(path) || bundleGraph.isAssetReferencedByAssetType(mod, 'js')`, since the root issue is that `isUnusedValue` doesn't work across bundles. (`isAssetReferencedByAssetType` should really be renamed `isAssetReferencedByAnotherBundleOfType`).

Reply:
Thanks for the feedback. I update the code as suggested, just running it for JS assets does make sense to me.

I can do the renaming in this PR as well if you like to :)  Just let me know.

Reply:
I just pushed the renaming as all tests passed locally. Feel free to revert in case you don't want to have it in here or is not what you expected.